README.md
# CausalGPS


| Resource    |  Github Actions      |  Code Coverage  |
| ----------  | -------------------- |-----------------|
| Platforms   | Windows, macOS, Linux|  codecov        |
| R CMD check | [![R build status](https://github.com/NSAPH-Software/CausalGPS/workflows/R-CMD-check/badge.svg?branch=develop)](https://github.com/NSAPH-Software/CausalGPS/actions) | [![codecov](https://codecov.io/gh/NSAPH-Software/CausalGPS/branch/develop/graph/badge.svg?token=97PCUXRGXH)](https://app.codecov.io/gh/NSAPH-Software/CausalGPS/) |

Matching on generalized propensity scores with continuous exposures

## Summary

`CausalGPS` is an R package that implements matching on generalized propensity scores with continuous exposures. The package introduces a novel approach for estimating causal effects using observational data in settings with continuous exposures, and a new framework for GPS caliper matching that jointly matches on both the estimated GPS and exposure levels to fully adjust for confounding bias.

## Installation

- Installing from source

```r
library("devtools")
install_github("NSAPH-Software/CausalGPS")
library("CausalGPS")
```

- Installing from CRAN 

```r
install.packages("CausalGPS")
```

- Setting up docker environment

Developing Docker image can be downloaded from Docker Hub. See more details in docker_singularity.

## Usage

Input parameters:

 **`w`** A data.frame comprised of two columns: one contains the observed exposure variable, and the other is labeled as 'id'. The column for the outcome variable can be assigned any name as per your requirements.   
 **`c`**  A data.frame of includes observed covariate variables. It should also consist of a column named 'id'.
 **`ci_appr`** The causal inference approach. Possible values are:   
   - "matching": Matching by GPS   
   - "weighting": Weighting by GPS   
 **`gps_density`** Model type which is used for estimating GPS value, including
 normal (default) and kernel.   
 **`use_cov_transform`** If TRUE, the function uses transformer to meet the
  covariate balance.   
 **`transformers`** A list of transformers. Each transformer should be a
 unary function. You can pass name of customized function in the quotes.   
 Available transformers:   
   - pow2: to the power of 2   
   - pow3: to the power of 3   
 **`bin_seq`** Sequence of w (treatment) to generate pseudo population. If
 NULL is passed the default value will be used, which is
 `seq(min(w)+delta_n/2,max(w), by=delta_n)`.   
 **`exposure_trim_qtls`** A numerical vector of two. Represents the trim quantile
 level. Both numbers should be in the range of [0,1] and in increasing order
 (default: c(0.01,0.99)).   
 **`gps_trim_qtls`** A numerical vector of two. Represents the trim quantile
 level. Both numbers should be in the range of [0,1] and in increasing order
 (default: c(0.01,0.99)). 
 **`params`** Includes list of params that is used internally. Unrelated
  parameters will be ignored.   
 **`sl_lib`**: A vector of prediction algorithms. 
 **`nthread`** An integer value that represents the number of threads to be
 used by internal packages.   
 **`...`**  Additional arguments passed to different models.

### Additional parameters   
#### Causal Inference Approach (`ci.appr`)   
 
 - if ci.appr = 'matching':   
   - *dist_measure*: Distance measuring function. Available options:   
     - l1: Manhattan distance matching   
   - *delta_n*: caliper parameter.   
   - *scale*: a specified scale parameter to control the relative weight that
  is attributed to the distance measures of the exposure versus the GPS.   
   - *covar_bl_method*: covariate balance method. Available options:   
      - 'absolute'   
   - *covar_bl_trs*: covariate balance threshold   
   - *covar_bl_trs_type*: covariate balance type (mean, median, maximal)   
   - *max_attempt*: maximum number of attempt to satisfy covariate balance.   
   - See [create_matching()] for more details about the parameters and default
   values.   
 - if ci.appr = 'weighting':   
   - *covar_bl_method*: Covariate balance method.   
   - *covar_bl_trs*: Covariate balance threshold   
   - *max_attempt*: Maximum number of attempt to satisfy covariate balance.
   
- Generating Pseudo Population

```r
pseudo_pop <- generate_pseudo_pop(w,
                                  c,
                                  ci_appr = "matching",
                                  gps_density = "normal",
                                  use_cov_transform = TRUE,
                                  transformers = list("pow2", "pow3"),
                                  sl_lib = c("m_xgboost"),
                                  params = list(xgb_nrounds = 50,
                                                xgb_max_depth = 6,
                                                xgb_eta = 0.3,
                                                xgb_min_child_weight = 1),
                                  nthread = 1,
                                  covar_bl_method = "absolute",
                                  covar_bl_trs = 0.1,
                                  exposure_trim_qtls = c(0.01,0.99),
                                  max_attempt = 1,
                                  dist_measure = "l1",
                                  delta_n = 1,
                                  scale = 1)

```
`matching_fn` is Manhattan distance matching approach. For prediction model we use [SuperLearner](https://github.com/ecpolley/SuperLearner) package. SuperLearner supports different machine learning methods and packages. `params` is a list of hyperparameters that users can pass to the third party libraries in the SuperLearner package. All hyperparameters go into the params list.  The prefixes are used to distinguished parameters for different libraries. The following table shows the external package names, their equivalent name that should be used in `sl_lib`, the prefixes that should be used for their hyperparameters in the `params` list, and available hyperparameters. 

| Package name | `sl_lib` name | prefix| available hyperparameters |
|:------------:|:-------------:|:-----:|:-------------------------:|
| [XGBoost](https://xgboost.readthedocs.io/en/latest/index.html)| `m_xgboost` | `xgb_`|  nrounds, eta, max_depth, min_child_weight |
| [ranger](https://cran.r-project.org/package=ranger) |`m_ranger`| `rgr_` | num.trees, write.forest, replace, verbose, family |

`nthread` is the number of available threads (cores). XGBoost needs OpenMP installed on the system to parallel the processing. `use_covariate_transform` activates transforming covariates in order to achieve covariate balance. Users can pass custom function name in a list to be included in the processing. At each iteration, which is set by the users using `max_attempt`, the column that provides the worst covariate balance will be transformed.  

- Estimating GPS

```r
data_with_gps <- estimate_gps(w,
                              c,
                              gps_density = "normal",
                              params = list(xgb_nrounds = 50,
                                            xgb_max_depth = 6,
                                            xgb_eta = 0.3,
                                            xgb_min_child_weight = 1),
                              nthread = 1,                                
                              sl_lib = c("m_xgboost")
                              )

```

- Generating Pseudo Population with an available GPS object

```r
gps_obj <- estimate_gps(w,
                        c,
                        gps_density = "normal",
                        params = list(xgb_max_depth = c(3,4,5),
                                      xgb_nrounds=c(10,20,30,40,50,60)),
                        nthread = 1,
                        sl_lib = c("m_xgboost")
                        )

pseudo_pop <- generate_pseudo_pop(w,
                                  c,
                                  ci_appr = "matching",
                                  gps_obj = gps_obj,
                                  use_cov_transform = TRUE,
                                  trim_quantiles = c(0.01,0.99),
                                  covar_bl_method = "absolute",
                                  covar_bl_trs = 0.1,
                                  covar_bl_trs_type = "mean",
                                  dist_measure = "l1",
                                  max_attempt = 1,
                                  delta_n = 1,
                                  scale = 0.5,
                                  nthread = 12)

```

- Estimating Exposure Rate Function

```r
estimate_npmetric_erf <- function(matched_Y,
                                  matched_w,
                                  matched_counter = NULL,
                                  bw_seq = seq(0.2,2,0.2),
                                  w_vals,
                                  nthread)
```

- Generating Synthetic Data

```r
syn_data <- generate_syn_data(sample_size=1000,
                              outcome_sd = 10,
                              gps_spec = 1,
                              cova_spec = 1)

```

## Contribution

For more information about reporting bugs and contribution, please read the contribution page from the package web page. 

## Code of Conduct

Please note that the CausalGPS project is released with a [Contributor Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct.html). By contributing to this project, you agree to abide by its terms.


## References

- CausalGPS method paper

```
@article{wu2022matching,
  title={Matching on generalized propensity scores with continuous exposures},
  author={Wu, Xiao and Mealli, Fabrizia and Kioumourtzoglou, Marianthi-Anna and Dominici, Francesca and Braun, Danielle},
  journal={Journal of the American Statistical Association},
  pages={1--29},
  year={2022},
  publisher={Taylor \& Francis}
}
```

- CausalGPS software paper

```
@misc{khoshnevis2023causalgps,
      title={CausalGPS: An R Package for Causal Inference With Continuous Exposures}, 
      author={Naeem Khoshnevis and Xiao Wu and Danielle Braun},
      year={2023},
      eprint={2310.00561},
      archivePrefix={arXiv},
      primaryClass={stat.CO},
      DOI={h10.48550/arXiv.2310.00561}
}
```



NEWS.md
## CausalGPS (Developing)


## CausalGPS 0.4.1 (2023-09-29)

### Added

* Extra step to check consistency of `delta_n` with exposure range.
* Software paper examples were added.

### Changed

* Plotting pseudo population includes object details. Set `include_details = TRUE`.  
*`generate_pseudo_pop` does not take `Y` as an input. 


## CausalGPS 0.4.0 (2023-05-25)

### Changed

* Docker image supports R 4.2.3
* `generate_syn_data` supports `vectorized_y` to accelerate data generation.
* `matching_fun` --> `dist_measure`
* `matching_l1` --> `matching_fn`
* `estimate_semipmetric_erf` now takes the `gam` models optional arguments.
* `estimate_pmetric_erf` now takes the `gnm` models optional arguments.
* `trim_quantiles` --> `exposure_trim_qtls`
* `generate_pseudo_pop` function accepts `gps_obj` as an optional input.
* `internal_use` is not part of parameters for `estimate_gps` function. 
* `estimate_gps` function only returns `id`, `w`, and computed `gps` as part of dataset.
* Now the design and analysis phases are explicitly separated.
* `gps_model` --> `gps_density`. Now it takes, `normal` and `kernel` options instead of `parametric` and `non-parametric` options.


### Added

* `estimate_npmetric_erf` supports both `locpol` and `KernSmooth` approaches.
* There is `gps_trim_qtls` input parameter to trim data samples based on gps values.
* Now users can also collect the original data in the pseudo population object.

### Fixed

* A bug with swapping transformed covairates with original one.

## CausalGPS 0.3.1 (2023-05-15)

### Changed

* Some of unit tests have less accuracy to overcome the bug with `stats::density` function. 

## CausalGPS 0.3.0 (2023-02-15)

### Changed

* Unit tests support new `wCorr` release (#193).
* Only optimized compilation is supported. In the previous versions, this approach is known as `optimzied_compile == TRUE`.

### Added

* The `earth` package is part of suggested packages. 


## CausalGPS 0.2.9 (2022-12-16)

### Fixed

* fixed a bug based on covariate balance threshold (#178, @naeemkh). 
* `estimate_npmetric_erf` assigns user-defined log file.

### Changed
* The process now prints the progress message based on the selected thresholds.
* In `estimate_npmetric_erf`:
  * `matched_Y` --> `m_Y`
  * `matched_w` --> `m_w`
  * `matched_cw` --> `counter_weight`
* In `estimate_npmetric_erf` function, the `matched_cw` input is now mandatory. 
* Internal kernel smoothing now uses `locpol::locpol` function.
* The entire data set is trimmed based on trimming quantiles.
* `earth` and `ranger` are not installed automatically. They can be installed manually if needed.
* `sysdata.rda` is modified to reflect transition from `counter` and `ipw` to `counter_weight`
* `counter_weight` is used as a counter or weight, in `matching` or `weighting`
approaches. `counter` and `ipw` are dropped.
* `sl_lib` becomes a required argument.
* The package has been transferred into NSAPH-Software Github account.
* Summary function of `gpsm_pspop` S3 object returns details of the adjusting process. 


### Added
* Now `Kolmogorov-Smirnov(KS)` statistics are provided for the computed pseudo population.
* `effect size` for the generated pseudo population is computed and reported.
* Binary search approach is used when scale = 1.
* `pseodo_pop` also includes covariate column names.
* `compute_closest_wgps_helper_no_sc` is added to take care of the mostly used special case (scale = 1).

### Removed 

* Dropped importing `KernSmooth` and `tidyr` packages. 
* `pred_model` argument dropped. The package only predicts using SuperLearner.


## CausalGPS 0.2.8 (2022-06-22)

### Fixed
* Message for not implemented methods changed to reduce misunderstanding.
* Empty counter will raise error in estimating non-parametric response function. 

### Changed
* matching_l1 returns frequency table instead of entire vector. 
* Vectorized population compilation and used data.table for multi-thread assignment.
* Removed nested parallelism in compiling pseudo population, which results in close control on memory.
* estimate_npmetric_erf also returns optimal h and risk values.

### Added
* `estimate_gps` returns the optimal hyperparameters.
* `estimate_gps` returns S3 object. 
* Internal xgboost approach support `verbose` parameter.
* Pseudo-population object now report the parameters that are used for the best covariate balance.


## CausalGPS 0.2.7 (2022-02-04)

### Fixed
* Naming covariate balance scores.

### Changed
* Restarting adaptive approach to keep trying up to maximum attempt.

### Added
* Synthetic data (synthetic_us_2010)
* Check on not defined covariate balance (absolute_corr_fun, absolute_weighted_corr_fun)
* Covariate balance threshold type: mean, median, maximal. 
* Improved test coverage.
* Singularity definition file.

## CausalGPS 0.2.6 (2021-09-06)

### Added
* added the status of optimized compile to generate_pseudo_pop function output.
* compute_closest_wgps accepts the number of user-defined threads.

### Changed

* Vignette file names.
* The trim condition from > and < into >= and <=.
* Removed seed input from generate_syn_data function. In R package, setting seed value inside function is not recommended. Users can set the seed before using the function. 
* OpenMP uses user defined number of cores.

### Fixed

* Initial covariate balance for weighted approach. The counter column was not preallocated correctly.
* Counter value for compiling. The initial value was set to one, which, however, zero is the correct one. 
* Private variable issue with OpenMP.
* Fixed OpenMP option on macOS checks. 

### Removed

## CausalGPS 0.2.5 (2021-07-23)

### Changed

* User needs to activate the logger

### Fixed

* CRAN package URLs are in canonical forms.

## CausalGPS 0.2.4 (2021-07-11)

### Added

* OpenMP for Rcpp code
* optimized_compile
* log_system_info()
* Frequently asked questions
* logo 

### Changed

* estimate_gps.Rmd
* estimate_semi_erf -> estimate_semipmetric_erf

R/plot.R
#' @title
#' A helper function for gpsm_erf object
#'
#' @description
#' A helper function to plot gpsm_erf object using ggplot2 package.
#'
#' @param object A gpsm_erf object.
#' @param ... Additional arguments passed to customize the plot.
#'
#' @return
#' Returns a ggplot object.
#'
#' @export
#'
#' @keywords internal
#' @importFrom ggplot2 autoplot
#' @importFrom rlang .data
#'
autoplot.gpsm_erf <- function(object, ...) {

  gg_labs <- gg_title <- NULL

  ## collect additional arguments
  dot_args <- list(...)
  arg_names <- names(dot_args)

  for (i in arg_names) {
    assign(i, unlist(dot_args[i], use.names = FALSE))
  }

  if (is.null(gg_labs)) {
    gg_labs <- c("Exposure Level", "Outcome Rate")
  }

  if (is.null(gg_title)) {
    gg_title <- "Exposure Response Curve"
  }

  # extract data
  tmp_data <- data.frame(w.vals = object$params$w_vals,
                         erf = object$erf)

  g <- ggplot2::ggplot(tmp_data) +
       ggplot2::theme_bw() +
       ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5)) +
       ggplot2::geom_line(ggplot2::aes(.data$w.vals, .data$erf),
                          color = "red", size = 1)


  g <- g + ggplot2::labs(x = gg_labs[1], y = gg_labs[2])
  g <- g + ggplot2::ggtitle(gg_title)

  return(g)
}

#' @title
#' Extend generic plot functions for gpsm_erf class
#'
#' @description
#' A wrapper function to extend generic plot functions for gpsm_erf class.
#'
#' @param x  A gpsm_erf object.
#' @param ... Additional arguments passed to customize the plot.
#'
#' @return
#' Returns a ggplot2 object, invisibly. This function is called for side effects.
#'
#' @export
#'
plot.gpsm_erf <- function(x, ...) {
  g <- ggplot2::autoplot(x, ...)
  print(g)
  invisible(g)
}


#' @title
#' A helper function for gpsm_pspop object
#'
#' @description
#' A helper function to plot gpsm_pspop object using ggplot2 package.
#'
#' @param object A gpsm_pspop object.
#' @param ... Additional arguments passed to customize the plot.
#'
#' @return
#' Returns a ggplot object.
#'
#' @export
#'
#' @keywords internal
#' @importFrom ggplot2 autoplot
#' @importFrom rlang .data
#'
autoplot.gpsm_pspop <- function(object, ...){

  gg_labs <- gg_title <- include_details <- NULL

  if (object$params$ci_appr == "matching"){
    default_gg_labs <- list(x = "Absolute Correlation", y = "Covariates")
  } else if (object$params$ci_appr == "weighting"){
    default_gg_labs <- list(x = "Absolute Weighted Correlation", y="Covariates")
  }

  default_gg_title <- "Covariate Balance Test"

  ## collect additional arguments
  dot_args <- list(...)
  arg_names <- names(dot_args)

  for (i in arg_names){
    assign(i,unlist(dot_args[i], use.names = FALSE))
  }

  # create a data.frame from two data.
  balance <- data.frame(original = object$original_corr_results$absolute_corr,
                        adjusted = object$adjusted_corr_results$absolute_corr)

  # Convert row names in to attribute.
  balance$covar_label <- row.names(balance)

  # sort data.frame based on original data correlation values
  balance <- balance[order(balance$original), ]
  covar_label <- balance$covar_label
  row.names(balance) <- NULL

  n_cov <- length(balance$original)
  m_balance <- melt(as.data.table(balance), measure.vars = c("original",
                                                             "adjusted"))
  m_balance$covariates <- rep(seq(1, n_cov, 1), 2)

  g <- ggplot2::ggplot(data = m_balance,
                       ggplot2::aes(x=.data$value,
                                    y=.data$covariates,
                                    color=.data$variable,
                                    shape=.data$variable)) + # Add shape aesthetic
    ggplot2::geom_point() +
    ggplot2::geom_path() +
    ggplot2::scale_color_manual(values = c("original" = "blue", "adjusted" = "orange")) + # Define colors manually
    ggplot2::scale_shape_manual(values=c(16, 15)) +  # Differentiate shapes: 16 is a solid circle, 24 is a triangle
    ggplot2::scale_y_discrete(limit = factor(1:n_cov),labels = covar_label) +
    ggplot2::theme_bw() +
    ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5)) +
    ggplot2::geom_vline(xintercept = object$params$covar_bl_trs) +
    ggplot2::geom_vline(xintercept = object$adjusted_corr_results$mean_absolute_corr,
                        linetype="dotdash") +
    ggplot2::labs(x = default_gg_labs$x, y = default_gg_labs$y) +
    ggplot2::ggtitle(default_gg_title)

  if (!is.null(gg_labs)) {
    g <- g + ggplot2::labs(x=gg_labs[1], y=gg_labs[2])
  }

  if (!is.null(gg_title)) {
    g <- g + ggplot2::ggtitle(gg_title)
  }

  if (!is.null(include_details)){
    if (!is.logical(include_details)){
      stop(paste("include_details should be logical value.",
                 "Current: ", include_details))
      }
  } else {
    include_details <- FALSE
  }


  # legend position
  if (include_details) {
    g <- g + ggplot2::theme(legend.position = c(0.8, 0.1))
  }

  # Object details
  if (include_details){
    object_sum <- "Object Summary "
    object_sum <- paste0(object_sum, "\n", "--------------")
    object_sum <- paste0(object_sum, "\n", "Approach: ",
                         object$params$ci_appr)
    if (object$params$ci_appr == "matching"){
      object_sum <- paste0(object_sum, "\n", " scale: ",
                           round(object$params$scale, 3))
      object_sum <- paste0(object_sum, "\n", " delta: ",
                           round(object$params$delta, 3))
      object_sum <- paste0(object_sum, "\n", " dist measure: ",
                           object$params$dist_measure)
    }
    object_sum <- paste0(object_sum, "\n", "exposure trim qtls: ",
                         round(object$exposure_trim_qtls[1],3), ", ",
                         round(object$exposure_trim_qtls[2],3), ")")
    object_sum <- paste0(object_sum, "\n", "gps trim qtls: ",
                         "(",
                         round(object$gps_trim_qtls[1],3), ", ",
                         round(object$gps_trim_qtls[2],3), ")")
    object_sum <- paste0(object_sum, "\n", "# sample (trimmed/org): ",
                         nrow(object$pseudo_pop), "/", object$original_data_size)
    object_sum <- paste0(object_sum, "\n", "--------------")
    object_sum <- paste0(object_sum, "\n", "Covar balance method: ",
                         object$params$covar_bl_method)
    object_sum <- paste0(object_sum, "\n", " Threshhold type: ",
                         object$params$covar_bl_trs_type)
    object_sum <- paste0(object_sum, "\n", " Threshhold value: ",
                         round(object$params$covar_bl_trs, 3))
    object_sum <- paste0(object_sum, "\n", " Passed covariate test: ",
                         object$passed_covar_test)
    object_sum <- paste0(object_sum, "\n", " Maximal abs. cov.: ",
                         round(object$adjusted_corr_results$maximal_absolute_corr, 3))
    object_sum <- paste0(object_sum, "\n", " Median abs. cov.: ",
                         round(object$adjusted_corr_results$median_absolute_corr, 3))
    object_sum <- paste0(object_sum, "\n", " Mean abs. cov.: ",
                         round(object$adjusted_corr_results$mean_absolute_corr, 3))
    object_sum <- paste0(object_sum, "\n", " Number of attemtps: ",
                         object$counter, "/",
                         object$params$max_attempt)
    object_sum <- paste0(object_sum, "\n", " Use cov transform: ",
                         object$use_cov_transform)
    object_sum <- paste0(object_sum, "\n", "--------------")
    object_sum <- paste0(object_sum, "\n", "Kolmogorov Smirnov Test")
    object_sum <- paste0(object_sum, "\n", " ess: ", round(object$ess,3))
    object_sum <- paste0(object_sum, "\n", " ess (min recomennded):", round(object$ess_recommended,3))
    n_gps_param <- length(object$best_gps_used_params)

    if (n_gps_param > 0) {
      gps_params <- object$best_gps_used_params
      object_sum <- paste0(object_sum, "\n", "--------------")
      object_sum <- paste0(object_sum, "\n", "SL params")
      for (item in seq(1,n_gps_param)){
        object_sum <- paste0(object_sum, "\n  ", names(gps_params[item]),": ",
                             round(gps_params[[item]],5))
      }
    }

    text_grob <- cowplot::ggdraw() + cowplot::draw_label(object_sum,
                                                         fontface = 'plain',
                                                         hjust = 0,
                                                         vjust = 1,
                                                         x = 0.01,
                                                         y = 0.95,
                                                         size = 7,
                                                         lineheight = 1.2)
  }

  if (include_details){
    g <- cowplot::plot_grid(
      g, text_grob,
      ncol = 2,
      rel_widths = c(2, 1)
    )
  }

  return(g)
}

#' @title
#' Extend generic plot functions for gpsm_erf class
#'
#' @description
#' A wrapper function to extend generic plot functions for gpsm_erf class.
#'
#' @param x  A gpsm_erf object.
#' @param ... Additional arguments passed to customize the plot.
#' @details
#' ## Additional parameters
#' - *include_details*: If set to TRUE, the plot will include run details (Default = FALSE).
#'
#' @return
#' Returns a ggplot2 object, invisibly. This function is called for side effects.
#'
#' @export
#'
plot.gpsm_pspop <- function(x, ...) {
  g <- ggplot2::autoplot(x, ...)
  print(g)
  invisible(g)
}
